### PR TITLE
FOGL-6366.v2 System Test to look for memory leaks

### PR DESCRIPTION
**Pull request by Mohit04tomar**
I have created the Jenkins job for memory leakage check using Valgrind.
Please check Jira for CI Job details.

### CODE REVIEW

**Review by MarkRiddoch — APPROVED** [review]

File: tests/system/memory_leak/scripts/setup [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# cloning plugins based on parameters passed to the script
+for i in ${1}
+do
+	echo $i
+	git clone  https://github.com/fledge-iot/${i}.git && cd ${i}; plugin_dir=`pwd`
+
+        # Cheking requirements.sh file exists or not, to install plugins dependencies
+	if [[ -f ${plugin_dir}/requirements.sh  ]]
+	then
+		./${plugin_dir}/requirements.sh
```
Comment (AmandeepArora) [inline tests/system/memory_leak/scripts/setup:63]: sudo is required here?

File: tests/system/memory_leak/scripts/setup [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	fi
+
+	# checking  CMakeLists.txt exists or not, to confirm whther it is a C based plugin or python based plugin
+	if [[ -f  ${plugin_dir}/CMakeLists.txt ]]
+	then
+		sed -i 's|c++11 -O3|c++11 -O0 -ggdb|g'  ${plugin_dir}/CMakeLists.txt
+		# building C based plugin
+		echo 'Building C plugin'
+		mkdir -p build && cd build && cmake -DFLEDGE_INSTALL=${FLEDGE_ROOT} -DFLEDGE_ROOT=${FLEDGE_ROOT} .. && make  && make install
```
Comment (AmandeepArora) [inline tests/system/memory_leak/scripts/setup:72]: This should be done within "${plugin_dir}"

File: tests/system/memory_leak/scripts/setup [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	else 
+		# Checking requirements.txt file exists or not, to install plugins dependency (if any)
+		if [[ -f ${plugin_dir}/requirements.txt  ]]
+		then
+			python3 -m pip install -r ${plugin_dir}/requirements.txt
+		fi
+		# checking whether plugin is south bound plugin or north bound plugin
+		echo 'Copying plugin'
+		if [[ $plugin_dir == *"south"* ]]
```
Comment (AmandeepArora) [inline tests/system/memory_leak/scripts/setup:81]: You can copy python plugins into FLEDGE_ROOT with this "sudo cp -r $plugin_dir/python $FLEDGE_ROOT/", irrespective of them being south/north/rule/delivery, etc.

File: tests/system/memory_leak/test_memcheck.sh [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+source config.sh
+
+export FLEDGE_ROOT=$(pwd)/fledge
+
+FLEDGE_TEST_BRANCH=$1     #here fledge_test_branch means branch of fledge repository that is needed to be scanned, default is devops
+
+cleanup(){
+  # sudo rm -rf /usr/local/fledge
+  rm -rf /tmp/*valgrind*.log /tmp/*valgrind*.xml
+  echo "Removig Repositories..."
```
Comment (AmandeepArora) [inline tests/system/memory_leak/test_memcheck.sh:14]: "Removing cloned repositories and temp files"?

File: tests/system/memory_leak/test_memcheck.sh [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+           "value": "true"
+        }
+     }
+  }'
+  echo
+  echo 'North setup done'
+}
+
+generate_valgrind_logs(){
+  sleep ${TEST_RUN_TIME}
```
Comment (AmandeepArora) [inline tests/system/memory_leak/test_memcheck.sh:87]: Sleep and fledge stop don't really belong to function named "generate_vangrind_logs". Move these 2 steps outside.

File: tests/system/memory_leak/scripts/setup (hunk 1 above)
Comment (Mohit04tomar) [reply tests/system/memory_leak/scripts/setup:63]: I have tried and run the script and requirements.sh file of fledge repository is not executing without using "sudo". So I did not use sudo in requirements.sh of plugin repository. 
One thing to note here is that the main script test_mem_check runs with sudo by Jenkins job, and that the main script calls setup script.

File: tests/system/memory_leak/scripts/setup (hunk 2 above)
Comment (Mohit04tomar) [reply tests/system/memory_leak/scripts/setup:72]: Yes, it is done within "${plugin_dir}"

File: tests/system/memory_leak/scripts/setup (hunk 1 above)
Comment (AmandeepArora) [reply tests/system/memory_leak/scripts/setup:63]: ok

File: tests/system/memory_leak/test_memcheck.sh (hunk 4 above)
Comment (AmandeepArora) [reply tests/system/memory_leak/test_memcheck.sh:14]: Sure. I was actually pointing to the typo here "Removig", so you may as well update the comment, if you want.

File: tests/system/memory_leak/scripts/setup (hunk 2 above)
Comment (AmandeepArora) [reply tests/system/memory_leak/scripts/setup:72]: ok

File: tests/system/memory_leak/test_memcheck.sh (hunk 5 above)
Comment (Mohit04tomar) [reply tests/system/memory_leak/test_memcheck.sh:87]: ok, I'll move it outside of that function

File: tests/system/memory_leak/scripts/setup (hunk 3 above)
Comment (Mohit04tomar) [reply tests/system/memory_leak/scripts/setup:81]: ok, I'll update it also in the next commit.

File: tests/system/memory_leak/test_memcheck.sh (hunk 4 above)
Comment (Mohit04tomar) [reply tests/system/memory_leak/test_memcheck.sh:14]: ok, I'm doing the changes

**Review by AmandeepArora — DISMISSED** [review]
Looks good

**Review by AmandeepArora — APPROVED** [review]

File: tests/system/memory_leak/test_memcheck.sh [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  mkdir -p reports/test1 ; ls -lrth
+  echo 'copying reports '
+  cp -rf /tmp/*valgrind*.log /tmp/*valgrind*.xml reports/test1/.  && echo 'copied'
+  rm -rf fledge*
+}
+
+cleanup
+setup
+reset_fledge
+adding_sinusoid
```
Comment (ashish-jabble) [inline tests/system/memory_leak/test_memcheck.sh:106]: `add_sinusoid`

**Pull request merged**